Added ToTimezoneIntervalFunction to handle Intervals in to_timezone

ganesh-vk opened the pull request:
This draft aims to add support for the interval type for the to_timezone function. Currently the case for (timezone.isConstant()) is incomplete and I'd appreciate some clarification regarding the same. 
- From what I gathered from the slack channel, I need to implement a getIntervalFunction with parameters similar to the getTimestampFunction present in ToUTCTimestampFunctionFactory, and handle both the low and high timestamps independently.
- How Can I get the Lo and Hi values of the interval since the interval is of type Function and I do not have a Record to pass as an argument. Also do I add this function to the same file? 

Please do let me know if there are any other changes that need to be made. Thank you!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=5089) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=5089) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=5089) it.</sub>

bluestreak01 commented:
hey @ganesh-vk thanks for the contribution! Do you fancy improving PR? In the following areas:

- code clean-up and sanity (warnings etc)
- reduce GC and allocations
- reduce time zone parsing overhead

Please let me know. If you do not have time - we can take this over.

ganesh-vk commented:
Hi, @bluestreak01 I've tried to reduce allocation however it does reduce code readability. If this was not what you had in mind please let me know! As for timezone parsing overhead and code clean up sanity, I did not quite understand and would appreciate a little more clarification.

ganesh-vk commented:
I've looked at #5286 and I assume the timezone parsing solution that needs to be implemented is similar to that implemented in the dateadd function if the timezone is constant? If so I can try implementing something similar.

bluestreak01 commented:
Hey @ganesh-vk i wanted to know if you are around. I will leave PR comments in an hour or so and give you ideas.

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    @NotNull
+    static IntervalFunction getIntervalFunction(IntList argPositions, Function interval, Function timezone, int multiplier) throws SqlException {
```
multiplier value is redundant, it is always 1 and i am not clear on the purpose of it

ganesh-vk replied:
This implementation was similar to the pre-existing implementation present in `to_timzone` and `to_utc` functions that deal with timestamps and not intervals. Currently `to_utc` does not support intervals either. When converting to UTC, the multiplier value is -1.

bluestreak01 replied:
the multiplier elsewhere is a code reuse hack to accommodates both millis and micros math, see `AbstractTimeZoneRules`. It always `a minute` in either millis or micros. What we're doing here is always in micros and factually not required.

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+            final long l = Timestamps.parseOffset(tz, 0, hi);
+            if (l == Long.MIN_VALUE) {
+                try {
+                    return new OffsetIntervalFunctionFromRules(
```
two functions can be consolidated into one if you use `Timestamps.getTimezoneRules` . This function will generate rules for all possible timezone variations

bluestreak01 replied:
additionally, all function class names are typically suffixed with `Function`

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/OffsetIntervalFunctionFromRules.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public Interval getInterval(Record rec) {
+        return new Interval(
```
all objects returned by Functions are expected to be mutable and reusable. `Interval` is also a reusable class. Do not create new objects for every value

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/OffsetIntervalFunctionFromRules.java:
```diff
@@ -0,0 +1,63 @@
+    public Interval getInterval(Record rec) {
+        return new Interval(
+                (interval.getInterval(rec).getLo() + multiplier * rules.getOffset(interval.getInterval(rec).getLo())),
+                (interval.getInterval(rec).getHi() + multiplier * rules.getOffset(interval.getInterval(rec).getHi()))
```
try not to invoke `getInterval()` multiple times. Function impls must not make assumptions of base function implementations and cache values always

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/OffsetIntervalFunctionFromRules.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public Interval getInterval(Record rec) {
+        return new Interval(
+                (interval.getInterval(rec).getLo() + multiplier * rules.getOffset(interval.getInterval(rec).getLo())),
```
readability wise, `interval` is an unfortunate name for a Function, perhaps use one of `base` or `intervalFunction`

bluestreak01 commented on core/src/test/java/io/questdb/test/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactoryTest.java:
```diff
@@ -0,0 +1,131 @@
+import io.questdb.std.str.StringSink;
+import io.questdb.std.Interval;
+
+public class ToTimezoneIntervalFunctionFactoryTest extends AbstractCairoTest {
```
it would be good to have a test where interval covers wither side of DST

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+        }
+
+        @Override
+        public Interval getInterval(Record rec) {
```
nit: @Nullable annotation is missing for the return value

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+        public Interval getInterval(Record rec) {
+            final Interval localInterval = new Interval(interval.getInterval(rec).getLo(), interval.getInterval(rec).getHi());
+            try {
+                final CharSequence tz = timezone.getStrA(rec);
```
null or invalid timezone must produce `CairoException`, which can also take SQL position to aid user fixing the issue. Functions have to behave consistently regardless if timezone constant value or not.

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+
+        @Override
+        public Interval getInterval(Record rec) {
+            final Interval localInterval = new Interval(interval.getInterval(rec).getLo(), interval.getInterval(rec).getHi());
```
reuse interval instance

ganesh-vk replied:
I have been trying to find an example where an Interval's value is altered, but to no avail. Am I supposed to declare a new instance variable of Type `Interval` and obtain it's timestamp values from `Function interval` and reuse that or is there a way to directly use the current `interval` variable of type `Function` , if so I'd appreciate some insight on how to do so. Thank you!

ganesh-vk replied:
I noticed that the current interval variable is of type Function, which prevents me from using .of directly on it. To work around this, I've introduced another variable of type Interval to achieve the desired functionality. Please let me know if this is not the intended approach. Thank you!

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+            try {
+                final CharSequence tz = timezone.getStrA(rec);
+                if (tz != null) {
+                    localInterval.of(
```
Use `Timestamps.getTimezoneRule` to lookup rule only once per value

bluestreak01 commented on core/src/main/java/io/questdb/griffin/engine/functions/date/ToTimezoneIntervalFunctionFactory.java:
```diff
@@ -0,0 +1,143 @@
+            return timezone;
+        }
+    }
+}
```
please add `toPlan(Sink)` function to serialize argument values and add test for plan (e.g. `explain select ...`)

ganesh-vk commented:
Hi, a quick update on the changes just made:

- Modified both versions (where timezone.isConstant() = true and timezone.isConstant = false) to use `Timestamps.getTimezoneRules` instead of parsing timezone/using Timestamps.toTimezone.
- Renamed `OffsetIntervalFunctionFromRules` to `OffsetIntervalFromRulesFunction`
- I have currently commented out the old getIntervalFunction and kept the OffsetIntervalFunctionFromOffset file, will remove both if these changes look good.
- Added daylight savings test.
- A previous test that used to return an interval in case of an invalid timezone now fails and returns an HTTP 400 (Bad Request) with the message "invalid timezone name." The test is testVarInvalidTimezone, which was copied from ToTimezoneTimestampFunctionFactoryTest, where the original function returns the input timestamp and does not throw an error. I am unsure whether the function is supposed to return the original interval in this case or throw an error. Could you please clarify how this should be handled? Thank you!
